src/main/scheduler/scheduler.c

```diff
@@ -350,79 +375,143 @@ static void readSchedulerLocals(task_t *selectedTask, uint8_t selectedTaskDynami
     task_t *selectedTask = NULL;
     uint16_t selectedTaskDynamicPriority = 0;
     uint16_t waitingTasks = 0;
-    bool realtimeTaskRan = false;
-    timeDelta_t gyroTaskDelayUs = 0;
+    uint32_t nextTargetCycles = 0;
+    int32_t schedLoopRemainingCycles;
 
     if (gyroEnabled) {
         // Realtime gyro/filtering/PID tasks get complete priority
         task_t *gyroTask = getTask(TASK_GYRO);
-        const timeUs_t gyroExecuteTimeUs = getPeriodCalculationBasis(gyroTask) + gyroTask->desiredPeriodUs;
-        gyroTaskDelayUs = cmpTimeUs(gyroExecuteTimeUs, currentTimeUs);  // time until the next expected gyro sample
-        if (gyroTaskDelayUs <= 0) {
-            taskExecutionTimeUs = schedulerExecuteTask(gyroTask, currentTimeUs);
+        static uint32_t lastTargetCycles = 0;
+        nowCycles = getCycleCounter();
+#if defined(UNIT_TEST)
+        lastTargetCycles = clockMicrosToCycles(gyroTask->lastExecutedAtUs);
+#else
+        if (lastTargetCycles == 0) {
+            lastTargetCycles = nowCycles;
+        }
+#endif
+        if (desiredPeriodCycles == 0) {
+            desiredPeriodCycles = (int32_t)clockMicrosToCycles((uint32_t)gyroTask->desiredPeriodUs);
+        }
+        nextTargetCycles = lastTargetCycles + desiredPeriodCycles;
+        schedLoopRemainingCycles = cmpTimeCycles(nextTargetCycles, nowCycles);
+
+        // Tune out the time lost between completing the last task execution and re-entering the scheduler
+        if ((schedLoopRemainingCycles < schedLoopStartMinCycles) &&
+            (schedLoopStartCycles < schedLoopStartMaxCycles)) {
+            schedLoopStartCycles += schedLoopStartDeltaUpCycles;
+        }
+
+        // Once close to the timing boundary, poll for it's arrival
+        if (schedLoopRemainingCycles < schedLoopStartCycles) {
+            // Record the interval between scheduling the gyro task
+            static int32_t lastRealtimeStartCycles = 0;
+
+            if (schedLoopStartCycles > schedLoopStartMinCycles) {
+                schedLoopStartCycles -= schedLoopStartDeltaDownCycles;
+            }
+#if !defined(UNIT_TEST)
+            while (schedLoopRemainingCycles > 0) {
+                nowCycles = getCycleCounter();
+                schedLoopRemainingCycles = cmpTimeCycles(nextTargetCycles, nowCycles);
+            }
+#endif
+            DEBUG_SET(DEBUG_SCHEDULER_DETERMINISM, 0, clockCyclesTo10thMicros(cmpTimeCycles(nowCycles, lastRealtimeStartCycles)));
+            lastRealtimeStartCycles = nowCycles;
+
+            timeUs_t currentTimeUs = clockCyclesToMicros(nowCycles);
+            taskExecutionTimeUs += schedulerExecuteTask(gyroTask, currentTimeUs);
             if (gyroFilterReady()) {
                 taskExecutionTimeUs += schedulerExecuteTask(getTask(TASK_FILTER), currentTimeUs);
             }
             if (pidLoopReady()) {
                 taskExecutionTimeUs += schedulerExecuteTask(getTask(TASK_PID), currentTimeUs);
             }
             currentTimeUs = micros();
```

Comment:
`currentTimeUs` is never used, and then overwritten on line 494.

Reply:
Fixed in https://github.com/betaflight/betaflight/pull/10573/commits/c7028a551ea7cfa8ccad612b53ddf05e51df1bee